docs/guides/1230-persistent-cache.md

```diff
@@ -0,0 +1,216 @@
+run and **cache will be lost**.
+
+In this page, we will see how to use `--cache-from` and `--cache-to` flags
+keep a permanent cache, from a local environment to GitHub action.
+
+## Ephemeral cache
+
+As an example, we will use a Dagger plan to build a go program.
+
+You can use any go project and the following snippet to build it in a
+standard way
+
+```cue
+package ci
+
+import (
+  "dagger.io/dagger"
+  "universe.dagger.io/go"
+)
+
+dagger.#Plan & {
+  // Retrieve source from current directory
+  // Input
+  client: filesystem: ".": read: {
+    contents: dagger.#FS
+    include: ["**/*.go", "go.mod", "go.sum"]
+  }
+
+  // Output
+  client: filesystem: "./output": write: {
+    contents: actions.build.output
+  }
+
+  actions: {
+    // Alias on source
+    _source: client.filesystem.".".read.contents
+
+    // Build go binary
+    build: go.#Build & {
+      source: _source
+    }
+  }
+}
+```
+
+To build go binary, just tip `dagger do build`, it should take some time to
+install dependencies, build binary and output it.
+
+Here's an example of a run
+
+```shell
+[✔] actions.build.container                                               338.0s
+[✔] client.filesystem.".".read                                              0.1s
+[✔] actions.build.container.export                                          0.1s
+[✔] client.filesystem."./output".write                                      0.2s
+```
+
+Indeed, Dagger has an ephemeral cache so if you rerun it, that shouldn't take
+that long.
+
+```shell
+[✔] actions.build.container                                                 2.9s
+[✔] client.filesystem.".".read                                              0.0s
+[✔] actions.build.container.export                                          0.0s
+[✔] client.filesystem."./output".write                                      0.1s
+```
+
+But if you stop Buildkit daemon and remove volume, cache will be lost and
+all actions will be executed again.
+
+:::info
+Now we have seen how ephemeral cache works, let's continue to understand how
+store cache in your local filesystem, so you can clean your docker engine without
+losing all your CI's cache.
+:::
+
+## Persistent cache in your local filesystem
+
+To store cache in your local filesystem, you don't need much effort : just
+add `--cache-from type=local,mode=max,dest=<output folder>` to `dagger do build`.
+
+:::tip
+Using `mode=max` argument will cache **all** layers from intermediate
+steps, with is really useful in the context of Dagger where you will have
+multiple steps to execute.
+:::
+
+Here's an example that exports cache to `storage`.
+
+```shell
+dagger do build --cache-to type=local,mode=max,dest=storage 
+# ...
+
+tree storage -L 1   
+storage
+├── blobs
+├── index.json
+└── ingest
+```
+
+A new directory has been created that contains cache the run
+
+:::caution
+If you run different action in the plan, don't forget to store cache in different
+destination, so it will be not overwritten.
```

Comment:
I think this is worth explaining in a bit more detail. Basically, our advice is that the directory cache is exported to should be namespaced by the action.

This is also unfortunate because there is no way overlapping cache between the actions can be de-duped. Between that and the performance issues noted by @aluzzardi [here](https://github.com/dagger/dagger/discussions/1875#discussioncomment-2445460), I almost wonder if we should even be emphasizing the use of local cache at all and instead just recommend registry cache. AFAIK, the use case for local cache usually involves exporting to some local dir that is then uploaded to a remote backend and later redownloaded for subsequent CI runs. Given the tradeoffs, I want to be careful we don't end up recommending something to our users that isn't actually beneficial to them.

I don't have all the context that lead to this doc, so if there's a pressing user need for local cache export documentation, then that's totally fine of course, but we should make it clear what performance tradeoffs there are IMO.

Reply:
Yeah we can still document local exports, but IMO opinion it would be better to start with registry at the top of the doc, use it for explaining the concepts and then mention local export as an option too after. This is subjective though.

I think this `:::caution` note should also be a bit more clear. Something like:
```
Local cache exports will overwrite anything already present in the directory, including any previous cache exports. Cache for distinct actions can be exported to different directories in order to not overwrite each other, but this may currently result in duplicated data if the two actions share any exported cache data in common.
```

Reply:
Fixed in https://github.com/dagger/dagger/pull/2035/commits/1651d359cfc25c6d69b505ae6103eeb90a06dc66